Segfault with input driving output directly

I get the following segfault during packing (see attached zip for test case). I think this is related to outputs being driven directly by inputs.

```
$ bash run.sh
....
Info: dbg1 feeds SB_IO qpi_in_buf[1], removing $nextpnr_obuf dbg1.
Info: dbg2 feeds SB_IO qpi_in_buf[0], removing $nextpnr_obuf dbg2.
Info: dbg3 feeds SB_IO qpi_io_buf[0], removing $nextpnr_obuf dbg3.
Info: dbg4 feeds SB_IO qpi_io_buf[1], removing $nextpnr_obuf dbg4.
Info: qpi_io2 feeds SB_IO qpi_io_buf[2], removing $nextpnr_iobuf qpi_io2.
run.sh: line 2: 10962 Segmentation fault      nextpnr-ice40 --freq 25 --up5k --asc mlaccel.asc --pcf mlaccel.pcf --json mlaccel.json
```

Running it in valgrind gives me a bunch of invalid read errors, then it exits using the following message during timing budgeting:

```
Info: Annotating ports with timing budgets for target frequency 25.00 MHz
ERROR: no timing info for port 'I' of cell type '$nextpnr_obuf'
```

This is with current git master (2298c89).

[nextpnr_segfault.zip](https://github.com/YosysHQ/nextpnr/files/2441649/nextpnr_segfault.zip)

See commit f4bfb2a in https://github.com/SymbioticEDA/mlaccel. (Sorry everyone reading along, it's not public yet. But @daveshah1 should have access.)

This is not exactly the same code as for the error above, and for me it fails with a slightly different error (bus error instead of segfault):

```
mlaccel/demo$ make mlaccel.bin 
nextpnr-ice40 --freq 25 --up5k --asc mlaccel.asc --pcf mlaccel.pcf --json mlaccel.json
...
Info: Annotating ports with timing budgets for target frequency 25.00 MHz
Makefile:58: recipe for target 'mlaccel.asc' failed
make: *** [mlaccel.asc] Bus error
```

But running it in valgrind shows "invalid read" errors at the same packing step as above.

--- a/ice40/pack.cc
+++ b/ice40/pack.cc
@@ -391,6 +391,8 @@ static bool is_nextpnr_iob(Context *ctx, CellInfo *cell)
 static void pack_io(Context *ctx)
 {
     std::unordered_set<IdString> packed_cells;
+    std::unordered_set<IdString> delete_nets;
+
     std::vector<std::unique_ptr<CellInfo>> new_cells;
     log_info("Packing IOs..\n");
 
@@ -410,14 +412,20 @@ static void pack_io(Context *ctx)
                 log_info("%s feeds SB_IO %s, removing %s %s.\n", ci->name.c_str(ctx), sb->name.c_str(ctx),
                          ci->type.c_str(ctx), ci->name.c_str(ctx));
                 NetInfo *net = sb->ports.at(ctx->id("PACKAGE_PIN")).net;
+                if (((ci->type == ctx->id("$nextpnr_ibuf") || ci->type == ctx->id("$nextpnr_iobuf")) &&
+                     net->users.size() > 1) ||
+                    (ci->type == ctx->id("$nextpnr_obuf") && (net->users.size() > 2 || net->driver.cell != nullptr)))
+                    log_error("PACKAGE_PIN of SB_IO '%s' connected to more than a single top level IO.\n",
+                              sb->name.c_str(ctx));
+
                 if (net != nullptr) {
-                    ctx->nets.erase(net->name);
+                    delete_nets.insert(net->name);
                     sb->ports.at(ctx->id("PACKAGE_PIN")).net = nullptr;
                 }
                 if (ci->type == ctx->id("$nextpnr_iobuf")) {
                     NetInfo *net2 = ci->ports.at(ctx->id("I")).net;
                     if (net2 != nullptr) {
-                        ctx->nets.erase(net2->name);
+                        delete_nets.insert(net2->name);
                     }
                 }
             } else {
@@ -434,6 +442,9 @@ static void pack_io(Context *ctx)
     }
     for (auto pcell : packed_cells) {
         ctx->cells.erase(pcell);
+    }
+    for (auto dnet : delete_nets) {
+        ctx->nets.erase(dnet);
     }
     for (auto &ncell : new_cells) {
         ctx->cells[ncell->name] = std::move(ncell);

